src/components/ResourceList/components/FilterControl/components/DateSelector/DateSelector.tsx

```diff
@@ -49,200 +36,296 @@ export enum DateFilterOption {
   OnOrAfter = 'on_or_after',
 }
 
-class DateSelector extends React.PureComponent<CombinedProps, State> {
-  state: State = {
-    datePickerMonth: this.now.getMonth(),
-    datePickerYear: this.now.getFullYear(),
-    initialConsumerFilterKey: this.props.filterKey,
-  };
+// eslint-disable-next-line react/display-name
+export const DateSelector = memo(function DateSelector({
+  filterValue,
+  filterKey,
+  filterMinKey,
+  filterMaxKey,
+  dateOptionType,
+  onFilterValueChange,
+  onFilterKeyChange,
+}: DateSelectorProps) {
+  const now = new Date();
+
+  const {translate} = useI18n();
```

Comment:
Existing code tends to do `const i18n = useI18n();` (my personal preference) or `const intl = useI18n();` (a hangover from before we renamed this) rather than destructure.

We should stick to one style throughout the codebase. Which one do you prefer?

Reply:
I would definitely prefer `i18n` or `intl`, the name clashing seems odd, best to just stick with one. However, we usually deconstruct whenever possible in our code. So I'm wondering if we should use `const {translate} = useI18n()` to be consistent?  I'm wondering if the reason we didn't originally deconstruct was that iirc web was used as a guideline and they can't deconstruct because they need access to the context, however, we don't but we just didn't deconstruct? 🤷‍♂ Either way as long as well consistent I'm ok with it.

Reply:
i18n is an instance of the I18n class, so I'm kinda surprised that we can even destructure it in the first place. I thought destructuring only worked on objects.

To reduce others being suprised by that I'd be tempted to avoid destructuring and stick with `const i18n = useI18n()`.

Reply:
> i18n is an instance of the I18n class, so I'm kinda surprised that we can even destructure it  in the first place. I thought destructuring only worked on objects.

In JavaScript, mostly everything is going to be an object, even regex since it's not primitive. Classes themselves are not objects, but functions (typeof class Test {} === 'function') however they create a instance of the function in an object (typeof new Test === 'object'). The problem with destructuring is that it'll remove context.

```js
class Test {
	constructor() {
  	this.num = 5;
  }

	func() {
  	return this.num;
  }

}

// All good
const t = new Test();
t.func();

// Losing context
const {func} = t;
func(); // Bad error ☹️
```

We don't use `this` within `I18n` so we're ok deconstructing. 

> To reduce others being suprised by that I'd be tempted to avoid destructuring and stick with const i18n = useI18n().

Two more points to support both sides of the argument.

Unless your playing with `i18n` you won't know it originated from a class.

If we don't deconstruct we'll be able to add features that do use `this` within having to refactor.

Reply:
Huh neat! It seems like [I18n's translate function already uses `this`](https://github.com/Shopify/polaris-react/blob/master/src/utilities/i18n/I18n.ts#L35) so by that logic this shouldn't even work now. And yet it seems to be manage. That still feels like it's more by luck than judgement though, so I'd still prefer not destructuring

Reply:
Oops looks like we are. It's working because translate is class property (not currently part of JavaScript but in the works)

```js
class Test {
  func = () => {}
}
```

Could compile down to something like 

```js
class Test {
  constructor() {
    this.func = () => {}
  }
}
```

Lamda functions lack certain contexts, including `this` and *usually* have the environment they were created in.

```js
class Test {
  func = () => { console.log(this) }
}

const {func} = new Test;
func(); // All ok
```

```js
class Test {
  func() { console.log(this) }
}

// Sorta gross alternative without using class properties still deconstructing
const test = new Test;
const {func} = test;
const boundFunc = func.bind(test);
boundFunc(); // All ok
```

Either way, I'm ok with not deconstructing our classes, it'll help future proof us and reduce the change of errors. We should reach out for other opinions as well.

Reply:
Talked it over with @dleroux and we're all on board with not deconstructing